src/libsync/propagatedownloadzsync.cpp

```diff
@@ -0,0 +1,320 @@
+    const QString &remotepath, const QString &localpath, const QString &localtmppath, QFile *device,
+    const QMap<QByteArray, QByteArray> &headers, const QByteArray &expectedEtagForResume,
+    const QByteArray &zsyncData, QObject *parent)
+    : GETJob(account, remotepath, parent)
+    , _device(device)
+    , _item(item)
+    , _propagator(propagator)
+    , _headers(headers)
+    , _expectedEtagForResume(expectedEtagForResume)
+    , _hasEmittedFinishedSignal(false)
+    , _zsyncData(zsyncData)
+    , _zs(NULL)
+    , _zr(NULL)
+    , _localpath(localpath)
+    , _localtmppath(localtmppath)
+    , _nrange(0)
+    , _zbyterange(NULL)
+    , _current(0)
+    , _pos(0)
+    , _received(0)
+{
+    _thread.setObjectName("GETFIleZsync thread");
+}
+
+void GETFileZsyncJob::startCurrentRange()
+{
+    qlonglong start = _zbyterange[_current*2];
+    qlonglong end = _zbyterange[(_current*2)+1];
+    _headers["Range"] = "bytes=" + QByteArray::number(start) + '-' + QByteArray::number(end);
+    qCDebug(lcZsyncGet) << path() << "HTTP GET with range" << _headers["Range"];
+
+    QNetworkRequest req;
+    for (QMap<QByteArray, QByteArray>::const_iterator it = _headers.begin(); it != _headers.end(); ++it) {
+        req.setRawHeader(it.key(), it.value());
+    }
+
+    req.setPriority(QNetworkRequest::LowPriority); // Long downloads must not block non-propagation jobs.
+
+    sendRequest("GET", makeDavUrl(path()), req);
+
+    reply()->setReadBufferSize(16 * 1024); // keep low so we can easier limit the bandwidth
+    qCDebug(lcZsyncGet) << _bandwidthManager << _bandwidthChoked << _bandwidthLimited;
+    if (_bandwidthManager) {
+        _bandwidthManager->registerDownloadJob(this);
+    }
+
+    if (reply()->error() != QNetworkReply::NoError) {
+        qCWarning(lcZsyncGet) << " Network error: " << errorString();
+    }
+
+    _pos = 0;
+
+    connect(reply(), &QIODevice::readyRead, this, &GETFileZsyncJob::slotReadyRead);
+    connect(reply(), &QNetworkReply::downloadProgress, this, &GETFileZsyncJob::slotDownloadProgress);
+    connect(reply(), &QNetworkReply::metaDataChanged, this, &GETFileZsyncJob::slotMetaDataChanged);
+    connect(this, &AbstractNetworkJob::networkActivity, account().data(), &Account::propagatorNetworkActivity);
+}
+
+bool GETFileZsyncJob::finished()
+{
+    if (reply()->bytesAvailable()) {
+        return false;
+    }
+
+    if (_current < _nrange - 1) {
+        _current++;
+        startCurrentRange();
+        return false;
+    }
+
+    if (!_hasEmittedFinishedSignal) {
+        emit finishedSignal();
+    }
+
+    free(_zbyterange);
+    zsync_end_receive(_zr);
+    zsync_complete(_zs);
+    zsync_end(_zs);
+
+    _hasEmittedFinishedSignal = true;
+
+    deleteLater();
+    return true; // discard
+}
+
+void GETFileZsyncJob::seedFinished(void* zs) {
+    _thread.quit();
+    _thread.wait();
+
+    _zs = (struct zsync_state *)zs;
+    {   /* And print how far we've progressed towards the target file */
```

Comment:
Yes it's called only once, `progress` in zsync code refers to how many blocks we have that match. If you have a local file that matches all the blocks in the metadata of the remote file then it would say 100% basically.

Reply:
The comment itself is just copy/paste from zsync. Can reword if that's what's putting you off.

Reply:
Thanks for the explanation, doesn't need changes.